sdk/src/account.rs

```diff
@@ -21,6 +21,159 @@ pub struct Account {
     pub rent_epoch: Epoch,
 }
 
+//pub type AccountNoData = Account;
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Clone, Default, AbiExample)]
+pub struct AccountNoData {
+    /// lamports in the account
+    pub lamports: u64,
+    /// data held in this account
+    #[serde(with = "serde_bytes")]
+    pub data: Vec<u8>, // will be: Arc<Vec<u8>>,
```

Comment:
Just another thought for exploration, what what the tradeoffs be if instead of a separate trait/struct for AccountNoData, we generalized the `data` field of `Account` to be a

`
enum AccountData {
     Owned(Vec<u8>),
     Shared(Arc<Vec<u8>>)
}
`
Would that reduce a lot of the needs to implement separate traits for both `AccountNoData` and `Account`, or is it just a lot of headache from having to replumb everything in the system?

Reply:
@carllin, I am under the impression that we cannot change the contents of struct Account. Certainly the ABI tests complain if I change it. I already ran into a few other structs that had to remain using Account. Perhaps we just can't change the serialization. I am unclear of the actual constraints. One thing I thought I understood from the discussion we had with @sakridge was that the clearest way forward was a new struct, currently proposed to be called AccountNoData. I am happy for feedback on another way forward.

Reply:
As part next of this topic on enum, I like the idea. I just talked with @Lichtso about serialization/deserialization of accounts for bpf execution, whether accounts are read-only or writeable. I think the new AccountNoData may at some point BE an enum similar to or maybe an extension of [LoadedAccount](https://github.com/solana-labs/solana/blob/ee621878b058ea1068f007e3261b2b0acae386be/runtime/src/accounts_db.rs#L185). Perhaps the representation in memory of an account after being written by a bpf program could be indexes into a slice that was mapped into the vm's address space. account.lamports() would read the enum and go to the slice to read out the memory or the data. Once we wrote to a store, we would free the slice. When we next loaded the account, it woudl begin life as an arc to an account struct that had an arc to the data. But the memory representation of the actual account data would be abstracted behind impl functions on the AccountNoData struct. Perhaps similar to the [AnAccount](https://github.com/solana-labs/solana/blob/dd5554112d0edfd2d83b22789718893afc4426b8/sdk/src/account.rs#L104) trait I've been working out and trying to avoid ;-).

Reply:
I talked with Carl. We think Account can't change.